Account for duplicate variables in printRelayQuery

ryancole opened the pull request:
This pull request relates to #784. I've adjusted the `createVariable` function, within `src\traversal\printRelayOSSQuery.js`, so that it first checks the existing variables before creating a new variable. This is to assist Relay when printing queries that contain overlapping arguments. (If the variable type and value is the same, do we need to continue creating variables with the same type and value? This PR adds the logic that answers that with a *no*.)

Without knowing what existing code is at my disposal within all of the FB libraries, I went ahead and just use a simple for loop over the existing keys, since the return state needs the key anyway.

I also created a test case that, prior to adjusting `createVariable`, failed. After adjusting `createVariable` the test passes, and no other existing tests seem to fail. Yay?

josephsavona commented on src/traversal/__tests__/printRelayOSSQuery-test.js:
```diff
@@ -190,6 +190,34 @@ describe('printRelayOSSQuery', () => {
+        env: enumValue,
+      });
+      const alias1 = query.getChildren()[0].getSerializationKey();
+      const alias2 = query.getChildren()[1].getSerializationKey();
```
The strictest option here is to verify the exact alias used - let's follow the example from line 375 and use `const alias = generateRQLFieldAlias('notifications.environment(WEB)')`, and then `${alias}: ...` for both fields below to make it clear that the alias should be identical.

ryancole replied:
Ah nice, alright. I had skimmed looking for how to generate those ids, and used the one that jumped out at me.

josephsavona commented on src/traversal/__tests__/printRelayOSSQuery-test.js:
```diff
@@ -190,6 +190,34 @@ describe('printRelayOSSQuery', () => {
+      expect(text).toEqualPrintedQuery(`
+        query FooQuery($environment_0:Environment) {
+          defaultSettings {
+            ${alias1}: notifications(environment:$environment_0),
```
nit: commas not required in GraphQL

josephsavona commented on src/traversal/printRelayOSSQuery.js:
```diff
@@ -301,6 +301,14 @@ function createVariable(
+    if (existingVariable.type === type && Object.is(existingVariable.value, value)) {
+      return existingId;
+    }
+  }
```
This will make each call to `printRelayQuery` O(n^2) with the number of variables in the query. At the moment we inline most argument values instead of creating variables, but if we ever changed that the use of a `for` loop here could be prohibitively expensive. 

Instead, how about changing the printer state's `variableMap` to a `Map` (to allow objects as keys since `value` is mixed), and changing it from `{[variableID: string]: {type,value}}` to `Map<value, {type, variableID}>`. Then the logic here can be roughly:

```javascript
const {variableMap} = printerState;
if (variableMap.has(value)) {
  return variableMap.get(value).variableID;
} else {
  const variableID = name + '_' + base62(printerState.variableCount++);
  variableMap.set(value, {type, variableID});
  return variableID;
}
```

The final printed variables can be created by iterating over `variableMap` to produce an object of variableID -> value.

josephsavona commented:
@ryancole Thanks for working on this! See feedback for how to make this even better :-)

ryancole commented:
@josephsavona Alright, I think I've addressed your feedback. The code now uses `Map { value => { type, variableID }}`. Changing it from the previous `Object { variableID: { type, value }}` required a few changes elsewhere, and also removed the usage of a couple imports.

ryancole commented on src/traversal/printRelayOSSQuery.js:
```diff
@@ -28,11 +26,11 @@ type PrinterState = {
   fragmentTexts: Array<string>;
   variableCount: number;
-  variableMap: {[variableID: string]: Variable};
+  variableMap: Map<mixed, Variable>;
```
I went ahead and use the existing Flow type, `Variable`, here. I don't particularly feel like the new contents of this `Variable` object (`type` and `variableID`) are properly described by the name Variable, but I'm not sure if changing this to something like `VariableDetails` or something is necessary. I just left it as-is. I've also never used Flow, so if this is flat out incorrect, I can change it. :)

ryancole commented on src/traversal/printRelayOSSQuery.js:
```diff
@@ -64,12 +62,13 @@ function printRelayOSSQuery(node: RelayQuery.Node): PrintedQuery {
     'printRelayOSSQuery(): Unsupported node type.'
   );
+  const variables = {};
+  for (const [key, value] of variableMap) {
```
The previously used `mapObject` function didn't work on a `Map`, so I'm just recreating the end result with this `for` logic. If you think this would be cleaner being in its own function, I can pull it out into its own function.

ryancole commented:
Finally, let me know if you'd like me to squash these commits. I don't have a working Flow binary, on Windows, that was compatible with the required version, so one of the commits ended up being to use TravisCI's Flow. :)

josephsavona commented:
nice, thanks for the update!

josephsavona commented:
@facebook-github-bot import
